### PR TITLE
Feature/seab 1637/collection entry version db (#3846)

### PR DESCRIPTION
**Pull request by garyluu**
For hotfix version of #3080 

Should compare it to https://github.com/dockstore/dockstore/pull/3846 and hopefully nothing weird happened. There were some cherry-pick issues

For whoever is doing the hotfix -> develop merge.  There is going to be a conflict with the migrations files.  Develop already has it in 1.10.0.  This has it in 1.9.0.  Prolly keep this one.

### CODE REVIEW

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 2
           

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/EntryVersion.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6417846&bid=20793325)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 2
           

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/EntryVersion.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6417846&bid=20793325)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java [hunk 1]
```diff
@@ -151,15 +155,15 @@ public void setDescription(String description) {
...
-    public void addEntry(Entry entry) {
-        this.entries.add(entry);
+    public void addEntry(Entry entry, Version version) {
+        this.entries.add(new EntryVersion(entry, version));
     }
 
-    public void removeEntry(Entry entry) {
-        this.entries.remove(entry);
+    public void removeEntry(Entry entry, Long versionId) {
+        this.entries.removeIf(entryVersion -> entryVersion.getEntry().getId() == (entry.getId()) && ((entryVersion.getVersion() != null ? entryVersion.getVersion().getId() == versionId : null == versionId)));
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java:166]: This looks re-usable
Might be worth storing in EntryVersion as a comparator like https://github.com/dockstore/dockstore/blob/e8b4a6760787c9890817ac9a44b53d98bcb31f7d/dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ObsoleteUrlFactory.java#L119 or 

https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/core/FileFormat.java#L48

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/CollectionEntry.java [hunk 2]
```diff
@@ -13,19 +13,41 @@
...
 
     public CollectionEntry(long id, Date dbUpdateDate, String entryTypeString, SourceControl sourceControl, String organization, String repository, String entryName)  {
         setEntryType(entryTypeString);
         setDbUpdateDate(dbUpdateDate);
         setId(id);
         setEntryPath(sourceControl.toString(), organization, repository, entryName);
+        setVersionName(null);
     }
 
+    @SuppressWarnings("checkstyle:ParameterNumber")
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/CollectionEntry.java:26]: Are these different constructors with different parameters null? 
Might be able to re-use, for example 
`CollectionEntry(long id, Date dbUpdateDate, String entryTypeString, SourceControl sourceControl, String organization, String repository, String entryName)`
can call 
`this(id, dbUpdateDate, entryTypeString, sourceControl, organization, repository, entryName, null))`

(Typos above probably)
Other example https://stackoverflow.com/questions/1182153/constructor-overloading-in-java-best-practice

**Review by denis-yuen — COMMENTED** [review]
Looking good

**Review by denis-yuen — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]
Approved, but if we're not doing a hotfix release and there's already other stuff on the hotbix branch, may be better not to merge.

**Comment by denis-yuen** [comment]
Sounds good, sorry @garyluu

**Pull request closed without merging**